### ISSUE TITLE
Question: how to get "real-time" errors in VSCode?

### ISSUE THREAD

**Issue by SHxKM** [typeddjango/django-stubs#134]
I'm open to moving this question to the respective/most-suitable VSCode repo, in case it doesn't belong here.

I have the following for my `workspace` settings file:

```json
"python.linting.mypyEnabled": true,
    "python.linting.mypyArgs": [
        "--config-file=${workspaceFolder}/mypy.ini"
    ],
```

This in my `mypy.ini` file:

```ini
[mypy]
plugins = mypy_django_plugin.main


warn_unused_ignores = True
strict_optional = True
check_untyped_defs = True
follow_imports = silent
ignore_missing_imports = True
show_column_numbers = True

[mypy.plugins.django-stubs]
django_settings_module = my_app.settings.dev
```

When I run `mypy` like so:

```
PYTHONPATH=${PYTHONPATH}:${PWD} mypy core/selectors.py
```

I get the correct errors, for example:

> error: Incompatible return value type (got "int", expected "Iterable[SomeModel]")

But this is not highlighted in my editor in real-time/after save. 

However, if I remove the plugin section from the `mypy.ini` file, I do get an error in the editor:

> Incompatible return value type (got "int", expected "Iterable[SomeModel]") mypy(error)

But if I don't have the plugin in `mypy.ini`, I see a lot of Django-related quirks that this plugin takes care of. So, anyone managed to get "realtime" type-checking in the VSCode editor with this plugin specifically?

**Comment by SHxKM** [comment]
Hmmm, after more than a month I’m still not able to get the correct behavior for VSCode. Are people just running the mypy command manually..? What am I missing?

**Comment by sobolevn** [comment]
Sorry, I do not have any experience in this field. Because yes, I do run `mypy` manually. 

Would be happy to know the solution / error reason.

**Comment by bluetech** [comment]
One thing that might cause this is that django-stubs fails to initialize django, specifically your `django_settings_module`, `my_app.settings.dev`, is not found in the `sys.path` of the plugin when it is initialized. The relevant code is

https://github.com/typeddjango/django-stubs/blob/7ba578f6b2e79782f45076fbb8f48eaca656cb23/mypy_django_plugin/django/context.py#L48-L70

AFAICS, vscode sets the current working directory to the workspace path, and the plugin tries to help by adding the cwd to the sys.path. But in some project layouts this might not be sufficient. To verify this, you can try to edit the file `mypy_django_plugin/django/context.py` in your venv, and add `sys.path.append('/some/path')` after the `getcwd()` line, where `/some/path` is the path which allows importing `my_app.settings.dev`.

**Comment by SHxKM** [comment]
@bluetech - Thanks for trying to assist. I was really hopeful about this, but unfortunately trying several paths did not help. I'm still getting this:

> ##########Linting Output - mypy##########
> Error constructing plugin instance of NewSemanalDjangoPlugin

**Comment by frnhr** [comment]
I've had reasonable success with this: https://marketplace.visualstudio.com/items?itemName=matangover.mypy

Seems to not parse all settings from `mypy.ini`, as I'm sometimes getting different errors than what running mypy directly reports, but they are mostly in agreement.

**Comment by oystein-beaufort** [comment]
I'm late to the party, and I don't know if this will work for you (I'm using `virtualenv`), but I've had success in using `mypy` with the `mypy_django_plugin` in VSCode by doing the following. It seems setting the `PYTHONPATH` to the directory containing the Django project and apps is what the `mypy_django`-plugin needs to find the Django project's settings file and setup the plugin correctly, like @bluetech mentioned.

In `env/bin/activate` (my virtualenv-activate script) I added:
```
export PYTHONPATH="src"  # ie. src contains my django project and apps
export DJANGO_SETTINGS_MODULE="myproject.settings"  # ie. my settings file is located at src/myproject.settings
```

In `.env` (VSCode's env file) I similarly added:
```
PYTHONPATH="src"  
DJANGO_SETTINGS_MODULE="myproject.settings"
```

(FYI: Adding `PYTHONPATH=${PYTHONPATH}:${PWD}` to the `.env`file like I've read in some other forums didn't work for me, so I had to explicitly set it to "src")

**Other setup:**
`mypy.ini`:
```
[mypy]
files = src
strict_optional = True
ignore_missing_imports = True
follow_imports = silent
show_column_numbers = True

plugins =
    mypy_django_plugin.main

[mypy.plugins.django-stubs]
django_settings_module = "myproject.settings"  
```

`.vscode/settings.json`:
```
{
    "editor.formatOnSave": true,
    "python.pythonPath": "env/bin/python",
    "python.linting.enabled": true,
    "python.linting.pylintEnabled": false,
    "python.linting.flake8Enabled": true,
    "python.linting.mypyEnabled": true,
    "python.linting.mypyPath": "env/bin/mypy",
}
```

Hope this helps :)

**Comment by AdrienLemaire** [comment]
coc.nvim user here (vscode plugin for vim).

I'm also getting this error
```
##########Linting Output - mypy##########
Error constructing plugin instance of NewSemanalDjangoPlugin
```

My mypy.ini file:
```ini
[mypy]
plugins =
    mypy_django_plugin.main

[mypy.plugins.django-stubs]
django_settings_module = "project.settings.dev"
```
I could verify that my WorkspaceFolders are correctly set up
```
:let g:WorkspaceFolders
g:WorkspaceFolders    ['/home/dori/Projects/Work/git_repo/project', '/home/dori/Projects/Work/git_repo']
```

If I comment out the django_settings_module line in mypy.ini, I get
```
/home/dori/Projects/Work/git_repo/project/mypy.ini:0: error: 'django_settings_module' is not set: setting is not provided
Found 1 error in 1 file (checked 1 source file)
```

If I comment out the mypy_django_plugin.main line from mypy.ini, I get mypy linting working fine (except without django-stubs).

This is confusing. 

EDIT: Actually the problem was explained in #329, my code was broken which is why mypy with django-stubs would break as well.

**Comment by nuschk** [comment]
FWIW, here's our setup, which works for sublime and vscode, with both, django and drf plugins.

First, `mypy.ini` lives in the `src/` directory, nothing fancy here, but we have to specify the settings module relative to `src/`:

```ini
[mypy]
files = .
python_version = 3.6
warn_return_any = True
warn_unused_configs = True
ignore_missing_imports = True
plugins =
    mypy_drf_plugin.main,
    mypy_django_plugin.main

[mypy.plugins.django-stubs]
django_settings_module = "<projectname>.settings.development"

```

The plan now is to call the `mypy` command from the `src` directory (hence, all django imports will work). Sublime is configured to change the working directory before calling mypy, via the projects file. Here's a snippet:

```json
{
    "build_systems": [...],
    "folders": [...],
    "settings":
    {
        ...
        "SublimeLinter.linters.mypy.working_dir": "${project_path}/src/"
    }
}
```

For vscode, we use a small wrapper script which we put in the `scripts/` directory:

```bash
project_root=$(dirname $(dirname $(realpath $0 )))
cd $project_root/src
exec mypy "$@"
```

Then, vscode can be configured to use the wrapper instead:

```json
{
    ...
    "python.linting.mypyPath": "scripts/mypy",
}
```

**Comment by kszmigiel** [comment]
@nuschk @oystein-beaufort Great hacks guys! I'm no VSC user myself, but I think it would be great to have such information in our readme. Any on you is up to contribute?

**Comment by nuschk** [comment]
@kszmigiel Yes, sure. But then again, isn't there an easier solution to this? 

The problem seems to be that `django-stubs` is running actual code (`django.setup()`), using mypy's cwd as import root. Wouldn't it be much more straight-forward to have a setting like `django_root_directory`in which we could specify the cwd for those?

So, in context.py, instead of this:

```python
        # add current directory to sys.path
        sys.path.append(os.getcwd())
```

We would have to add the django root as well and it might just work (didn't try it though). The upside is, mypy could be called from anywhere, as it's supposed to be.

Another option would be to consider `MYPYPATH` for this, but that doesn't seem right to me.

**Comment by dmartin** [comment]
It seems to be possible to get this working now without any extra workaround, since https://github.com/typeddjango/django-stubs/commit/92eb068a040a1388e01e7f34da3f7a8542dacb47.

If you specify a `mypy_path` in the mypy config file, it'll include it when looking for `django_settings_module`.

**Comment by nuschk** [comment]
> It seems to be possible to get this working now without any extra workaround, since [92eb068](https://github.com/typeddjango/django-stubs/commit/92eb068a040a1388e01e7f34da3f7a8542dacb47).
> 
> If you specify a `mypy_path` in the mypy config file, it'll include it when looking for `django_settings_module`.

Can confirm it works for me now. 

I also use `mypy_drf_plugin`, which still has the problem. But, if you load the `mypy_django_plugin` first, even that works.

mypy.ini:

```ini
[...]
plugins =
    mypy_django_plugin.main,
    mypy_drf_plugin.main
```

**Comment by UnknownPlatypus** [comment]
Looking at the two last comments, this issue seems resolved!
